File: 2022/july/09-07-2022/jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        dp=[nums[0]]+[0]*(len(nums)-1)
        for i in range(1,len(nums)): dp[i]=nums[i]+max(dp[max(0,i-k):i])
        return dp[-1]

    def maxResult1(self, nums: List[int], k: int) -> int:
        heap=[(0,-k)]
        for i in range(len(nums)):
            while i-heap[0][1]>k: heappop(heap)
            nums[i]-=heap[0][0]
            heappush(heap,(-nums[i],i))
        return nums[-1]

File: 2022/july/09-07-2022/jump-game-vi.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn max_result(nums: Vec<i32>, k: i32) -> i32 {
		// score[i] = max { score[i-k], ..., score[i-1] } + nums[i]
        // maintain a priority queue to reduce each score calc to O(logn) time
        // time: O(nlogn)
        // space: O(n)
        let n = nums.len();
        let k = k as usize;
        let mut score = nums[0];
        let mut pq = BinaryHeap::new();
        pq.push((nums[0], 0));
        for i in 1..n {
            while let Some(&(v, idx)) = pq.peek() {
                if idx + k < i {
                    pq.pop();
                    continue;
                }
                score = v + nums[i];
                pq.push((score, i));
                break;
            }
        }
        score
    }
}
